crates/proof-of-sql/src/sql/parse/provable_expr_plan_builder.rs

```diff
@@ -100,9 +101,21 @@ impl ProvableExprPlanBuilder<'_> {
                 s.clone(),
                 s.into(),
             )))),
-            Literal::Timestamp(its) => Ok(ProvableExprPlan::new_literal(
-                LiteralValue::TimeStampTZ(its.timeunit, its.timezone, its.timestamp.timestamp()),
-            )),
+            Literal::Timestamp(its) => {
+                let timestamp = its.timestamp.timestamp_nanos_opt().ok_or_else(|| {
+                    PoSQLTimestampError::UnsupportedPrecision(
+                        "Timestamp out of range: 
+                    Valid nanosecond timestamps must be between 1677-09-21T00:12:43.145224192 
+                    and 2262-04-11T23:47:16.854775807."
+                            .to_owned(),
+                    )
+                })?;
+
+                Ok(ProvableExprPlan::new_literal(LiteralValue::TimeStampTZ(
+                    its.timezone,
+                    timestamp,
+                )))
+            }
```

Comment:
this was the cause of the bug. keeping the time unit would mean matching to each instance of it in the its (intermediate timestamp), but the intermediate timestamp has no way of knowing the precision of the column it will eventually be matched to just from this point